src/ray/object_manager/object_buffer_pool.h

```diff
@@ -0,0 +1,193 @@
+  std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status> GetChunk(const ObjectID &object_id, uint64_t data_size,
+                            uint64_t metadata_size, uint64_t chunk_index);
+
+  /// When a chunk is done being used as part of a get, this method is invoked
+  /// to indicate that the buffer associated with a chunk is no longer needed.
+  ///
+  /// \param object_id The object_id of the buffer to release.
+  /// \param chunk_index The index of the chunk.
+  /// \return The status of invoking this method.
+  ray::Status ReleaseGetChunk(const ObjectID &object_id, uint64_t chunk_index);
+
+  /// Returns a chunk of an empty object at the given chunk_index. The object chunk
+  /// serves as the data that is to be written to by a connection receiving an object
+  /// from a remote node. Only one chunk can be referenced at a time.
+  ///
+  /// \param object_id The ObjectID.
+  /// \param data_size The sum of the object size and metadata size.
+  /// \param metadata_size The size of the metadata.
+  /// \param chunk_index The index of the chunk.
+  /// \return A ChunkInfo struct.
+  std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status> CreateChunk(const ObjectID &object_id, uint64_t data_size,
+                               uint64_t metadata_size, uint64_t chunk_index);
+
+  /// Abort the create operation associated with a chunk at chunk_index.
+  /// This should not be called if SealChunk is called.
+  ///
+  /// \param object_id The ObjectID.
+  /// \param chunk_index The index of the chunk.
+  /// \return The status of invoking this method.
+  ray::Status AbortCreateChunk(const ObjectID &object_id, uint64_t chunk_index);
+
+  /// Seal the object associated with a create operation. This is invoked whenever
+  /// a chunk is successfully written to. This is not called if AbortCreateChunk
+  /// is called.
+  ///
+  /// \param object_id The ObjectID.
+  /// \param chunk_index The index of the chunk.
+  /// \return The status of invoking this method.
+  ray::Status SealChunk(const ObjectID &object_id, uint64_t chunk_index);
+
+ private:
+
+  /// Abort the create operation associated with an object. This destroys the buffer
+  /// state, including create operations in progress for all chunks of the object.
+  ///
+  /// \param object_id The ObjectID.
+  /// \return The status of invoking this method.
+  ray::Status AbortCreate(const ObjectID &object_id);
+
+  /// Abort the get operation associated with an object.
+  ray::Status AbortGet(const ObjectID &object_id);
+
+  /// Builds the chunk vector for an object, and store it by object_id in chunk_info_.
+  /// Returns the number of chunks into which object is split.
+  std::vector<ChunkInfo> BuildChunks(const ObjectID &object_id,
+                                     uint8_t *data,
+                                     uint64_t data_size);
+
+  /// Holds the state of a get buffer.
+  struct GetBufferState {
+    GetBufferState() {}
```

Comment:
Why the empty constructor?

Reply:
The compiler won't allow `unordered_map` with values that don't have a default constructor (so non-existent keys can be accessed).